storage: evaluate proposals without lock

tbg opened the pull request:
The main objective of this change is to (later) enable proposer-evaluated KV to
compute the `WriteBatch` when evaluating a proposal (i.e. when turning
a `BatchRequest` into a `*ProposalData`). The previous code assumes making
a proposal is cheap, and thus it is done under a large critical sections.

MaxLeaseIndex is assigned at the latest possible moment before submitting to
Raft. Multiple inbound requests can pass the command queue without interacting,
and it's important to assign the Lease index only right before when the
proposal is submitted to Raft.

Note that the command queue currently doesn't cover all the data accessed by
Raft commands (#10084) and that a discussion of how precisely command
evaluation will work is still being discussed (#6290). Nevertheless, this
change should be safe and a step in the right direction.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10012)

<!-- Reviewable:end -->

bdarnell commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KUDN8n5uGZP-Y8qVcaB:b-n9m6xo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
> ```Go
> // into Replica's proposal map and subsequently passed to submitProposalLocked.
> //
> // Replica.mu must not be held.
> ```

To be clear, it's no longer synchronized by Replica.mu, but it should only be called from the scheduler (which acts as a kind of mutex), right?

---

*[pkg/storage/replica.go, line 1682 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDBKeuLHWUBcBtkYJA:-KUDBKeuLHWUBcBtkYJB:b1ks4vm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1682)):*
> ```Go
> 	r.mu.Lock()
> 	if r.mu.destroyed != nil {
> 		return nil, nil, r.mu.destroyed
> ```

Missing `r.mu.Unlock()`

---

*[pkg/storage/replica.go, line 2302 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDCSEEItkMDqcKzDnV:-KUDCSEEItkMDqcKzDnW:b-ax1nqv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2302)):*
> ```Go
> // refreshPendingCmds with the ElectionTimeoutTicks as parameter.
> func (r *Replica) queueRefreshStalePendingCmds(reason refreshRaftReason) {
> 	log.Warningf(r.ctx, "queueing refresh of stale pending commands: %s", reason)
> ```

Are these warnings temporary?

---

*[pkg/storage/replica.go, line 2313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDM1qMh4Kt_HyVqUxX:-KUDM1qMh4Kt_HyVqUxY:b-bbhhv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2313)):*
> ```Go
> }
> 
> // refreshPendingCmds goes through the pending proposals, refurbishing what
> ```

These names are (still) really difficult to remember/distinguish. Does the move to proposer-evaluated KV suggest any better names? For example, could we now s/refurbish/reevaluate/?

---

*[pkg/storage/replica.go, line 2316 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDERrhDbUZ1LxIo9wD:-KUDERrhDbUZ1LxIo9wE:bcdorpp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2316)):*
> ```Go
> // it can and reproposing the rest. refreshAtDelta specifies how old (in ticks)
> // a command must be for it to be inspected; usually this is called with zero (affect everything) or the number of ticks of an election timeout (affect only
> // proposals that have had ample time to apply but didn't).
> ```

Proposer-evaluated KV makes it much more expensive to refresh a command, so we may need to be more conservative about how often we do it. 

---

*[pkg/storage/replica.go, line 2361 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDEh4fvRrNuNp4KPMn:-KUDEh4fvRrNuNp4KPMo:brdtp8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2361)):*
> ```Go
> 		if err := r.submitProposalLocked(p); err != nil {
> 			log.Warning(p.ctx, err)
> 			// TODO(tschottdorf): should notify the client and remove from
> ```

This will be fixed as a part of this PR, right?

---

*[pkg/storage/replica.go, line 2370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDENPidSA4OJoUmjP4:-KUDENPidSA4OJoUmjP5:bi0sjp6) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2370)):*
> ```Go
> 	r.mu.Unlock()
> 
> 	for _, p := range refurbish {
> ```

This reverses the order of repropsals and refurbishments. Why?

---

*[pkg/storage/replica.go, line 2671 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDGdbyBkiUbWWC3X4I:-KUDGdbzePHNiExA3JD5:bebw2tv) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2671)):*
> ```Go
> 				// refurbish the command. We do this asynchronously because
> 				// it's expensive and shouldn't block this goroutine.
> 				cmdCopy := *cmd // take copy so we can bend the done channel below
> ```

"bend"?

---

*[pkg/storage/replica.go, line 2688 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDH5bXPTeQ8kL7rBFS:-KUDH5bXPTeQ8kL7rBFT:b-ywzca4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2688)):*
> ```Go
> 						}
> 					}); pErr != nil {
> 						log.Info(r.ctx, pErr)
> ```

Include more context in this error message (or silently drop it? Does it matter that we give up on refurbishments during shutdown?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KUDR-Rs_PGOhNNnN7DZ:b2ajj7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> To be clear, it's no longer synchronized by Replica.mu, but it should only be called from the scheduler (which acts as a kind of mutex), right?
</details>
`Replica.Send` which calls down to `Replica.propose` is not run on the scheduler (unless something changed in this PR).

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KUDS4XjJSKgKl7qQbmw:biezrv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
> ```Go
> 	}
> 	if refreshReason != noReason {
> 		r.queueRefreshAllPendingCmds(refreshReason)
> ```

Why are you queueing the refresh vs executing it here? You're already running on a `raftScheduler` worker. 

---

*[pkg/storage/replica.go, line 2107 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUE8wfH1fjvVvXUC5qJ:-KUE8wfH1fjvVvXUC5qK:bningbf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2107)):*
> ```Go
> 		// through. In effect, proposals will will be refreshed when they have
> 		// been pending for 1 to 2 election cycles.
> 		r.queueRefreshStalePendingCmds(reasonTicks)
> ```

Similar to my other comment, you're already running on a `raftScheduler` worker goroutine here. Why are you enqueuing which will just cause another scheduler loop before processing the refresh?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KUHI5FYbgU-x-PuxDNo:b-aeco2e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `Replica.Send` which calls down to `Replica.propose` is not run on the scheduler (unless something changed in this PR).
</details>
The scheduler isn't involved here at all. I don't think this was ever spelled out anywhere (so we should do it in this change, especially now that guarantees are relaxed), but here's my understanding:

The replica mutex does not at all serialize proposals, but the command queue serializes proposals which may "conflict" in some manner. For simple key-based commands, it makes intuitive sense what that means (we should double-check what can go on with commands which skip the command queue, and look at leases, splits, ... once more).
Essentially, anywhere between `beginCmd` and the proposal applying, you can interleave arbitrarily with other "non-conflicting" commands, and they can pass through evaluation and various critical sections in some order (including reproposals, refurbishments, ...).
That needs to be correct in order for this change to be correct (and also for the old code to have been correct).
It also needs to be efficient, and I (just) changed one thing that could have bitten us, namely we now assign the lease counter IDs (MaxLeaseIndex) just before proposing to Raft. We used to do it when creating the command, but that is before the (soon to be) long evaluate section.

---

*[pkg/storage/replica.go, line 1682 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDBKeuLHWUBcBtkYJA:-KUHL5ik0J655K3mU7b1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1682)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Missing `r.mu.Unlock()`
</details>
Done.

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KUHPBAYWa3lYf4vWzJJ:b-u5e05t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why are you queueing the refresh vs executing it here? You're already running on a `raftScheduler` worker. 
</details>
I agree, but only because we're doing it at the end of the work (for example, it would seem wrong to let one command pay for the latency that comes with evaluating a potentially large number of proposals anew). For `tick` the same is true. I can change those two, though I think it would be simpler if there were one straightforward and safe way to make these calls (i.e. pay the scheduler round trip).

The more interesting invocation is that in `setReplicaID`, which is called (with some indirection) from `getOrCreateReplica`, so that is less tied to a goroutine (though in practice we only call it from two places, both of which are - again - on the scheduler). We definitely don't want to insert arbitrary delays there as those calls are early in the invocation.

And then there's the awkward one of the bunch, in which we can't go through the scheduler due to only wanting to target one specific command. How do you feel about that goroutine? Now that we have the scheduler, I feel pretty bad about offloading work in an uncontrolled way. But then again, we want some of that work done in parallel with standard Raft processing - by occupying the scheduler with reproposals, am I not also delaying the handling of "regular" Raft processing (since no other worker will process outstanding requests for a Range, which btw isn't documented).

So ISTM that I may not be doing performance a favor here by using the scheduler, and that I should *always* delegate?

---

*[pkg/storage/replica.go, line 2313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDM1qMh4Kt_HyVqUxX:-KUHLNOAtqiZZDwVv0qp:bcj6ii1) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2313)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> These names are (still) really difficult to remember/distinguish. Does the move to proposer-evaluated KV suggest any better names? For example, could we now s/refurbish/reevaluate/?
</details>
Yes, especially `refreshPendingCmds` should already have been renamed. I like `s/refurbish/reevaluate`, though I'd prefer doing another rename-only PR and keep this one as small as possible. I added TODOs for both renames.

---

*[pkg/storage/replica.go, line 2316 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDERrhDbUZ1LxIo9wD:-KUHLwV2LgvzSP4HaZme:bk67l0q) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2316)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Proposer-evaluated KV makes it much more expensive to refresh a command, so we may need to be more conservative about how often we do it. 
</details>
I agree. Anything you want me to do here?

---

*[pkg/storage/replica.go, line 2361 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDEh4fvRrNuNp4KPMn:-KUHOmgUhV7taG58OjSp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2361)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This will be fixed as a part of this PR, right?
</details>
Done.

---

*[pkg/storage/replica.go, line 2370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDENPidSA4OJoUmjP4:-KUHMMll25ilALoNV1zo:b-ol8eeu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2370)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This reverses the order of repropsals and refurbishments. Why?
</details>
Concretely because of what makes sense with the critical sections here, but also because I think that's the right order. Added a comment, PTAL and LMK if I'm confusing something there.

---

*[pkg/storage/replica.go, line 2671 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDGdbyBkiUbWWC3X4I:-KUHOzSxBITIOAdQZRFp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2671)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> "bend"?
</details>
Done.

---

*[pkg/storage/replica.go, line 2688 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDH5bXPTeQ8kL7rBFS:-KUHD1odfAVUpYP4XLbm:b-5mfehk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2688)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Include more context in this error message (or silently drop it? Does it matter that we give up on refurbishments during shutdown?)
</details>
In general it shouldn't matter that we don't refurbish, though I can think of at least one situation in which it leads to deadlock (a client is waiting on its done channel at shutdown, and can't abandon the command as it is being applied. So it won't be able to quit unless the command applies successfully, but if that needs a refurbishment, it will never do so).
I'm now sending the error directly to the client to avoid that ever happening.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 49 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 2302 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDCSEEItkMDqcKzDnV:-KUHUeT7V_4Rlo0984e3:b6npoi) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2302)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Are these warnings temporary?
</details>
I'll clean these up when the way forward becomes clear.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Addressed most comments, but some discussion remains. I changed the base branch of this PR, so please excuse the weird diffs Reviewable might throw at you.

---

Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-:-KUHUws3NphLecltwMWJ:b-ksx3w5)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 3 of 5 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KUJEoR4VBo2LwB_7zH2:bdfzvgf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The scheduler isn't involved here at all. I don't think this was ever spelled out anywhere (so we should do it in this change, especially now that guarantees are relaxed), but here's my understanding:
> 
> The replica mutex does not at all serialize proposals, but the command queue serializes proposals which may "conflict" in some manner. For simple key-based commands, it makes intuitive sense what that means (we should double-check what can go on with commands which skip the command queue, and look at leases, splits, ... once more).
> Essentially, anywhere between `beginCmd` and the proposal applying, you can interleave arbitrarily with other "non-conflicting" commands, and they can pass through evaluation and various critical sections in some order (including reproposals, refurbishments, ...).
> That needs to be correct in order for this change to be correct (and also for the old code to have been correct).
> It also needs to be efficient, and I (just) changed one thing that could have bitten us, namely we now assign the lease counter IDs (MaxLeaseIndex) just before proposing to Raft. We used to do it when creating the command, but that is before the (soon to be) long evaluate section.
</details>
You should convert some of the above into a comment in the code describing the correctness requirements. Something I'm not clear on is when a proposal needs to be re-evaluated vs re-proposed. With proposer-evaluated KV, can we ever simply re-propose a Raft command with reevaluating it?

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KUKBhX1WEa6-RiwStMc:bl69ati) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
> And then there's the awkward one of the bunch, in which we can't go through the scheduler due to only wanting to target one specific command. How do you feel about that goroutine? 

Can you elaborate on this? I'm not quite sure what you're referring to. 

Ok, if refreshing pending commands is going to be expensive, then I can see the value in not imposing that cost on the current operation being performed. But see my other comment about needing to clarify when reevaluation is necessary. Are there scenarios (perhaps common) where we can re-propose instead of reevaluate (and then re-propose)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 1 of 5 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KULh_AhwMHxlciBtd7k:b1owr9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
> You should convert some of the above into a comment in the code describing the correctness requirements.

Done (`addWriteCmd`).

> Something I'm not clear on is when a proposal needs to be re-evaluated vs re-proposed

This is documented inside of `refreshPendingCmds`, but not very well. I added it more explicitly:

> // Refurbishing roughly equates being allowed to assume that the previously
// submitted command can no more be applied successfully (due to the applied
// lease index having surpassed the proposal's MaxLeaseIndex). Reproposing
// refers to simply taking the in-flight proposal and submitting it to Raft
// again; this is what we have to do for everything that couldn't be
// refurbished as refurbishing anyway would lead to doubly-applying some of
// the proposals.

On your other question:

> With proposer-evaluated KV, can we ever simply re-propose a Raft command with reevaluating it?

Nothing there should have changed. Two commands which are in flight (i.e. past command queue) concurrently should not be able to influence each other. Of course that is one of the big assumptions that we should thoroughly try to poke holes into.

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KULhBzzbvgXxn227IjK:b4rfebs) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
> Can you elaborate on this? I'm not quite sure what you're referring to.

Pinged you from a comment there.

Could you get back to me on this:

> But then again, we want some of that work done in parallel with standard Raft processing - by occupying the scheduler with reproposals, am I not also delaying the handling of "regular" Raft processing (since no other worker will process outstanding requests for a Range, which btw isn't documented).

It's my biggest concern right now. Doesn't seem right to use the scheduler.

---

*[pkg/storage/replica.go, line 2687 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KULgxxC5w0hLkJzFi3M:-KULgxxC5w0hLkJzFi3N:bcczpz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/16c896e20a91e9ce691ecb71e4e0496352e0d79a/pkg/storage/replica.go#L2687)):*
> ```Go
> 				defer func() {
> 					if err := r.store.Stopper().RunAsyncTask(ctx, func(ctx context.Context) {
> 						if pErr := r.refurbishPendingCmd(&cmdCopy); pErr != nil {
> ```

@petermattis this one does part of the work that `refreshPendingCmds` does, but only for a single command.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1597 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDN8n5uGZP-Y8qVcaA:-KUMLoOX5acB9TXtwRgZ:b-bgikdp) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L1597)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > You should convert some of the above into a comment in the code describing the correctness requirements.
> 
> Done (`addWriteCmd`).
> 
> > Something I'm not clear on is when a proposal needs to be re-evaluated vs re-proposed
> 
> This is documented inside of `refreshPendingCmds`, but not very well. I added it more explicitly:
> 
> > // Refurbishing roughly equates being allowed to assume that the previously
> // submitted command can no more be applied successfully (due to the applied
> // lease index having surpassed the proposal's MaxLeaseIndex). Reproposing
> // refers to simply taking the in-flight proposal and submitting it to Raft
> // again; this is what we have to do for everything that couldn't be
> // refurbished as refurbishing anyway would lead to doubly-applying some of
> // the proposals.
> 
> On your other question:
> 
> > With proposer-evaluated KV, can we ever simply re-propose a Raft command with reevaluating it?
> 
> Nothing there should have changed. Two commands which are in flight (i.e. past command queue) concurrently should not be able to influence each other. Of course that is one of the big assumptions that we should thoroughly try to poke holes into.
</details>
The new comment looks good, thanks.

---

*[pkg/storage/replica.go, line 2313 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDM1qMh4Kt_HyVqUxX:-KUMNCDEVz5pOoMjNVaK:b-5ialu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2313)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yes, especially `refreshPendingCmds` should already have been renamed. I like `s/refurbish/reevaluate`, though I'd prefer doing another rename-only PR and keep this one as small as possible. I added TODOs for both renames.
</details>
Yeah, fine to do the renaming separately. 

---

*[pkg/storage/replica.go, line 2316 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDERrhDbUZ1LxIo9wD:-KUMNRd9RxIczTcl7SZg:bq8kfv4) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2316)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> I agree. Anything you want me to do here?
</details>
Nothing specific; just pointing it out. The only tuneable variable here is how many ticks a command must be outstanding to repropose, but I don't think it makes sense to extend that beyond an election timeout unless and until we see real problems. 

---

*[pkg/storage/replica.go, line 2370 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDENPidSA4OJoUmjP4:-KUMO-q6_hK0LMYXHOjG:b8uqcav) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2370)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Concretely because of what makes sense with the critical sections here, but also because I think that's the right order. Added a comment, PTAL and LMK if I'm confusing something there.
</details>
Yeah, makes sense.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KUYcnk4ZRFwTxVj0iYW:b2zvyqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> > Can you elaborate on this? I'm not quite sure what you're referring to.
> 
> Pinged you from a comment there.
> 
> Could you get back to me on this:
> 
> > But then again, we want some of that work done in parallel with standard Raft processing - by occupying the scheduler with reproposals, am I not also delaying the handling of "regular" Raft processing (since no other worker will process outstanding requests for a Range, which btw isn't documented).
> 
> It's my biggest concern right now. Doesn't seem right to use the scheduler.
</details>
`processRaftCommand` is currently serialized on Ranges. As discussed earlier today, it would be great to eventually parallelize processing within a Range, but I don't think proposer-evaluated KV should be gated on that. 

---

*[pkg/storage/replica.go, line 2687 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KULgxxC5w0hLkJzFi3M:-KUYbyzN5OaCcINbzb1R:b-xhmbkz) ([raw file](https://github.com/cockroachdb/cockroach/blob/16c896e20a91e9ce691ecb71e4e0496352e0d79a/pkg/storage/replica.go#L2687)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> @petermattis this one does part of the work that `refreshPendingCmds` does, but only for a single command.
</details>
Using a Raft scheduler worker wouldn't be much better than using the current goroutine (which is a Raft scheduler worker). My preference would be to do the refurbishment inline here with a TODO to investigate the benefit of using a separate goroutine. 

---

*[pkg/storage/replica.go, line 1531 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUY_JfTHITW1McjrpoO:-KUY_JfTHITW1McjrpoP:bkkkplc) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1531)):*
> ```Go
> // reorderings. For example, a batch of writes and a split could be in flight
> // in parallel without overlap, but if the writes hit the RHS, something must
> // prevent them from writing outside of the key range when they apply.
> ```

Per #10084, a split should lock-out the RHS of the split. I suppose you'll be updating this comment in that PR.

---

*[pkg/storage/replica.go, line 1532 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUY__mz2m8LGX5VoVuz:-KUY__mz2m8LGX5VoVv-:byyokh8) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1532)):*
> ```Go
> // in parallel without overlap, but if the writes hit the RHS, something must
> // prevent them from writing outside of the key range when they apply.
> // Similarly, a command proposed under lease A must not apply under lease B.
> ```

I'm not clear on what check you're going to perform to prevent this. We can't check the lease validity when Raft has committed the proposal and is asking for it to be applied as Raft can't depend on the wall time. Perhaps what you meant is `a command evaluated under lease A must not be proposed under lease B`. Or I could just be confused.

---

*[pkg/storage/replica.go, line 1697 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYaUhYc_HnNdsrrGAJ:-KUYaUhYc_HnNdsrrGAK:bfweiyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1697)):*
> ```Go
> 	// has a tiny (~1%) performance hit for single-node block_writer testing.
> 	r.raftMu.Lock()
> 	defer r.raftMu.Unlock()
> ```

I think you could delay this locking to just before `r.evaluateProposal`. 

---

*[pkg/storage/replica.go, line 2371 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYazf7Fpo_Evuvoaeo:-KUYazf7Fpo_Evuvoaep:b-glpd39) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L2371)):*
> ```Go
> // can and reproposing the rest.
> //
> // Refurbishing roughly equates being allowed to assume that the previously
> ```

`s/equates being/equates to being/g`

---

*[pkg/storage/replica.go, line 2379 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYbAYpYQdK6bmdCKh_:-KUYbAYpYQdK6bmdCKha:b-8jck87) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L2379)):*
> ```Go
> // the proposals.
> //
> // refreshAtDelta specifies how old (in ticks)
> ```

Nit: rewrap this paragraph.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I want to get https://github.com/cockroachdb/cockroach/pull/10131 in first. It will remove some of the complications here.

tschottdorf commented:
PTAL at the new changes (fair bit of churn between the revisions due to reworking this after #10131, perhaps just review from scratch).

---

Review status: 0 of 2 files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 2043 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUDS4XjJSKgKl7qQbmv:-KUqNg9IyGyJpmPDnngB:bfn1zo) ([raw file](https://github.com/cockroachdb/cockroach/blob/1a01c28c164b3435bf7202eccce517ad2a676da3/pkg/storage/replica.go#L2043)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `processRaftCommand` is currently serialized on Ranges. As discussed earlier today, it would be great to eventually parallelize processing within a Range, but I don't think proposer-evaluated KV should be gated on that. 
</details>
Thanks to #10131 (which removed refurbishments), this question has been resolved (by not needing to be addressed in the first place).

---

*[pkg/storage/replica.go, line 2687 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KULgxxC5w0hLkJzFi3M:-KUqO3QNH1NYVLak65Zx:bvpssbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/16c896e20a91e9ce691ecb71e4e0496352e0d79a/pkg/storage/replica.go#L2687)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Using a Raft scheduler worker wouldn't be much better than using the current goroutine (which is a Raft scheduler worker). My preference would be to do the refurbishment inline here with a TODO to investigate the benefit of using a separate goroutine. 
</details>
Absent in newer revision.

---

*[pkg/storage/replica.go, line 1531 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUY_JfTHITW1McjrpoO:-KUqM81lnL1-fP869Okl:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1531)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Per #10084, a split should lock-out the RHS of the split. I suppose you'll be updating this comment in that PR.
</details>
Yep.

---

*[pkg/storage/replica.go, line 1532 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUY__mz2m8LGX5VoVuz:-KUqMBYNVjnCKdBtQA4o:bw693dg) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1532)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'm not clear on what check you're going to perform to prevent this. We can't check the lease validity when Raft has committed the proposal and is asking for it to be applied as Raft can't depend on the wall time. Perhaps what you meant is `a command evaluated under lease A must not be proposed under lease B`. Or I could just be confused.
</details>
No, it's meant as written. Checking downstream of Raft is done today and is always possible. It only depends on the command's timestamp and the lease under which it commits. Why I'm pointing it out here is because we need more than what's currently done (for scenarios in which lease A and lease B are both held by the same node but the node restarted in between and lost its command queue, for example. That would allow scenarios in which a write pops up out of nowhere and applies successfully, invalidating reads which were served).

---

*[pkg/storage/replica.go, line 1697 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYaUhYc_HnNdsrrGAJ:-KUqNdYAtysVLj5YXwqD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L1697)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I think you could delay this locking to just before `r.evaluateProposal`. 
</details>
Done.

---

*[pkg/storage/replica.go, line 2371 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYazf7Fpo_Evuvoaeo:-KUqNvrEw9gHsAPDXKKi:bvpssbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L2371)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> `s/equates being/equates to being/g`
</details>
Absent in newer revision.

---

*[pkg/storage/replica.go, line 2379 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUYbAYpYQdK6bmdCKh_:-KUqO0OEmTHnLaKAoca3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/50770af24c98323a90cd3306758753f95f1010c8/pkg/storage/replica.go#L2379)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Nit: rewrap this paragraph.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-:-KUqO6VBMx14raoJ5JrB:b-cx1ons)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Have you run `make stress PKG=./storage`? 

---

Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 294 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUrxmDOgSMtgG8HzUB5:-KUrxmDOgSMtgG8HzUB6:brx1p6f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L294)):*
> ```Go
> 		// The *ProposalData in the map are "owned" by it. Elements from the
> 		// map must only be referenced while Replica.mu is held, except if the
> 		// element is removed from the map first.
> ```

Did you audit the code to make sure this is true?

---

*[pkg/storage/replica.go, line 296 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUrxkaxbAQKzz0d1UBQ:-KUrxkaxbAQKzz0d1UBR:bif77gw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L296)):*
> ```Go
> 		// element is removed from the map first.
> 		//
> 		// TODO(tschottdorf): evaluate whether this should be a list/slice.
> ```

Is this `TODO` still valid?

---

*[pkg/storage/replica.go, line 1743 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUryL8xUggZHAlibrkc:-KUryL8xUggZHAlibrkd:b-tls40h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L1743)):*
> ```Go
> 	//
> 	// TODO(tschottdorf): holding raftMu during evaluateProposal may not be the
> 	// right thing once we actually do evaluation work there. See #10084.
> ```

I'd rephrase this to say that holding `raftMu` during evaluation will limit concurrency at the range level and is something we will eventually need to address.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-:-KUrz2eIiz68FHZaT9PV:b-1lqa31)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: evaluate proposals without lock

The main objective of this change is to (later) enable proposer-evaluated KV to
compute the `WriteBatch` when evaluating a proposal (i.e. when turning
a `BatchRequest` into a `*ProposalData`). The previous code assumes making
a proposal is cheap, and thus it is done under a large critical sections.

MaxLeaseIndex is assigned at the latest possible moment before submitting to
Raft. Multiple inbound requests can pass the command queue without interacting,
and it's important to assign the Lease index only right before when the
proposal is submitted to Raft.

Note that the command queue currently doesn't cover all the data accessed by
Raft commands (#10084) and that a discussion of how precisely command
evaluation will work is still being discussed (#6290). Nevertheless, this
change should be safe and a step in the right direction.

tschottdorf commented:
TFTR. `make stress` is running: `866 runs so far, 0 failures, over 10m50s`. I'll leave it on for a few more hours.

---

Review status: 0 of 2 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 294 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUrxmDOgSMtgG8HzUB5:-KUs6D4Zx7TRLaHtRAy0:bp5peil) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L294)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Did you audit the code to make sure this is true?
</details>
Yes. This was mostly relevant during refurbishment, which has now changed to deleting from the map and notifying the client, in that order (which is also the order in which it always happens, as audited).

---

*[pkg/storage/replica.go, line 296 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUrxkaxbAQKzz0d1UBQ:-KUs6L4ndUy_6qJSJ_bA:bm75cpj) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Is this `TODO` still valid?
</details>
Yes, but it's in an odd place and has no context. Moved to `refreshProposalsLocked` where it belongs.

---

*[pkg/storage/replica.go, line 1743 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUryL8xUggZHAlibrkc:-KUs6qQXoc_1e2RoGeN8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d5beaf3d04072db250ae0800511fac12f29a9b90/pkg/storage/replica.go#L1743)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I'd rephrase this to say that holding `raftMu` during evaluation will limit concurrency at the range level and is something we will eventually need to address.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-:-KUs6s9Emf4oxi2NbcGw:b45w7yp)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Did get some stress failures, but nothing that seems related:

```
--- FAIL: TestStoreScanIntents (0.12s)
       	store_test.go:1814: 0: expected scan count 2; got 3

--- FAIL: TestRaftRemoveRace (0.05s)
        client_test.go:926: [n1,s1,r1/1:/M{in-ax}]: change replicas aborted due to failed
 preemptive snapshot: range=1: remote declined snapshot:
```

The later is tracked in #10166, but I'll have to double check on the former.

tschottdorf commented:
Indeed seems newly flaky:

`10241 runs completed, 1 failures, over 1m36s`

(ran on master for >9min)

I'll take a look tomorrow.

tschottdorf commented:
Upon further testing, it's also flaky on master. Filed https://github.com/cockroachdb/cockroach/issues/10198.

tamird commented:
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2500 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdEhb_nmXZWRSA856:-KUwdEhct-Ej334MvjWG:b-jnb72r) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica.go#L2500)):*
> ```Go
> 	// definitely required, however.
> 	//
> 	// TODO(tschottdorf): evaluate whether `proposals` should be a list/slice.
> ```

`proposals` meaning `r.mu.proposals`?

---

*[pkg/storage/replica_test.go, line 5789 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdP17p70jtwZLn6fT:-KUwdP17p70jtwZLn6fU:b-2dtycw) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5789)):*
> ```Go
> 	// Disable ticks to avoid automatic reproposals after a timeout, which
> 	// would pass this test.
> 	cfg.RaftTickInterval = time.Hour
> ```

If you're disabling ticks, why not use math.MaxInt64? `time.Hour` invites speculation.

---

*[pkg/storage/replica_test.go, line 5932 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdgDHZ_mT3jdWQjL0:-KUwdgDHZ_mT3jdWQjL1:b-ycyeuo) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5932)):*
> ```Go
> 	const num = 10
> 
> 	var chs []chan proposalResult
> ```

you might want a comment here that explains `chs` is protected by `rng.mu`.

---

*[pkg/storage/replica_test.go, line 5953 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdYhg4Ak-Bs8g1EvC:-KUwdYhg4Ak-Bs8g1EvD:bi7bndc) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5953)):*
> ```Go
> 				t.Error(err)
> 			} else {
> ```

NYC but random whitespace here =/

---

*[pkg/storage/replica_test.go, line 6065 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdrdoKZXXcR6wy6rm:-KUwdrdoKZXXcR6wy6rn:bfyt2a0) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L6065)):*
> ```Go
> 	var tc testContext
> 	cfg := TestStoreConfig()
> 	cfg.RaftTickInterval = 24 * time.Hour // disable ticks
> ```

again, yet this time you're using 24 hours? this invites speculation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 2500 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdEhb_nmXZWRSA856:-KV0ehEYzEJd9WTPpF0T:bb77ucv) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica.go#L2500)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `proposals` meaning `r.mu.proposals`?
</details>
Yep.

---

*[pkg/storage/replica_test.go, line 5789 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdP17p70jtwZLn6fT:-KV0eia2qr3YksWz9llG:bg5nc5d) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5789)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> If you're disabling ticks, why not use math.MaxInt64? `time.Hour` invites speculation.
</details>
No good reason.

---

*[pkg/storage/replica_test.go, line 5932 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdgDHZ_mT3jdWQjL0:-KV0ek-FVMXFe30MzNs3:bb73djm) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5932)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you might want a comment here that explains `chs` is protected by `rng.mu`.
</details>
SGTM

---

*[pkg/storage/replica_test.go, line 5953 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdYhg4Ak-Bs8g1EvC:-KV0eqkU2GOvdclp5rOw:bhvj20m) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L5953)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> NYC but random whitespace here =/
</details>
Kill it! :)

---

*[pkg/storage/replica_test.go, line 6065 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012#-KUwdrdoKZXXcR6wy6rm:-KV0eod5NygQSVC7nRE_:bdujrz) ([raw file](https://github.com/cockroachdb/cockroach/blob/41526e2569ce25715e2422df7640a8d3c686d942/pkg/storage/replica_test.go#L6065)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> again, yet this time you're using 24 hours? this invites speculation.
</details>
MaxInt64 SGTM

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10012)*
<!-- Sent from Reviewable.io -->
